book/scheduling-and-threading.md

```diff
@@ -278,11 +277,48 @@ class JSContext:
                 Task(self.interp.evaljs,
                     "__runSetTimeout({})".format(handle)))
 
-        set_timeout(run_callback, time / 1000.0)
+        threading.Timer(time / 1000.0, run_callback).start()
 ```
 
 That's it! Your browser now supports running asynchronous JavaScript tasks.
 
+Except there is a bug in this code: it doesn't account for the fact that
+the first thread and the timer thread run concurrently, and there is therefore
+no guarantee that one callback that writes to `schedule_task` will not be 
+interleaved with code on the other thread trying to read the task queue,
+leading to a [race condition](https://en.wikipedia.org/wiki/Race_condition)
+bug and nondeterministic results.
+
+This bug is easily fixed by use of a `threading.Lock` object. Before reading
+or writing to a data structure shared across threads, acquire the lock; after
+you're done, release it.^[The `blocking` parameter to `acquire` indicates
+whether the thread should block on acquiring the lock or not; in this chapter
+you'll always set it to true.] The code changes in `TaskRunner` are pretty
+easy---just be careful to not forget to release the lock, and hold it for the
+minimum time possible, so as to maximize thread parallelism.^[That's why the
+code releases the lock before calling `task`: after the task has been removed
+from the queue, it can't be accessed by another thread.]
```

Comment:
I think the point about releasing before calling `task` is good and shouldn't be confined to a footnote.

Reply:
Done.